pkg/safepath/join_linux.go

```diff
@@ -0,0 +1,204 @@
+//
+// When the returned path is consumed by the caller, it is his responsibility
+// to call the cleanup function which will close the descriptor.
+func safeOpenFd(path, subpath string) (string, func(), error) {
+	noCleanup := func() {}
+
+	// Open base volume path (_data directory).
+	prevFd, err := unix.Open(path, unix.O_PATH, 0)
+	if err != nil {
+		return "", noCleanup, ErrorNotAccessible{}
+	}
+
+	// Try to use the Openat2 syscall first (available on Linux 5.6+).
+	fd, err := unix.Openat2(prevFd, subpath, &unix.OpenHow{
+		Flags:   unix.O_PATH,
+		Mode:    0,
+		Resolve: uint64(unix.RESOLVE_BENEATH | unix.RESOLVE_NO_MAGICLINKS),
+	})
+
+	// Openat2 is available and succeeded.
+	if err == nil {
+		unix.Close(prevFd)
+		fdPath := procFsFdPath(fd)
+		return fdPath, func() {
+			if err := unix.Close(fd); err != nil {
+				logrus.WithError(err).Warn("failed to close fd")
+			}
+		}, nil
+	}
+
+	// Openat2 is available and failed, return error.
+	if err != nil && !errors.Is(err, unix.ENOSYS) {
+		unix.Close(prevFd)
+
+		if errors.Is(err, unix.EXDEV) {
+			return "", noCleanup, ErrorEscapesBase{}
+		} else if errors.Is(err, unix.ENOENT) {
+			return "", noCleanup, ErrorNotAccessible{}
+		}
+		return "", noCleanup, errors.Wrap(err, "openat2 failed")
+	}
+
+	// Openat2 is not available, fallback to Openat loop.
+	fdPath, fd, err := softOpenat2(prevFd, subpath)
+	if err != nil {
+		unix.Close(prevFd)
+		return "", noCleanup, err
+	}
+
+	return fdPath, func() {
+		if err := unix.Close(fd); err != nil {
+			logrus.WithError(err).Warn("failed to close fd")
+		}
+	}, nil
+}
+
+func softOpenat2(baseFd int, subpath string) (string, int, error) {
+	basePath, mountSource, err := getFdRealpath(baseFd)
+	if err != nil {
+		unix.Close(baseFd)
+		return "", -1, err
+	}
+
+	prevFd := baseFd
+	parts := strings.Split(subpath, string(os.PathSeparator))
+	for _, part := range parts {
+		// Open next path part relative to the previous descriptor.
+		fd, err := unix.Openat(prevFd, part, unix.O_PATH, 0)
+		if err != nil {
+			unix.Close(prevFd)
+			return "", -1, ErrorNotAccessible{}
+		}
+
+		// Close previous base fd and set the new one as a base for the next Openat.
+		if err = unix.Close(prevFd); err != nil {
+			logrus.WithError(err).WithField("part", part).Warn("error closing previous fd")
+		}
+		prevFd = fd
+
+		// Check for path escape.
+		realPath, procPath, err := getFdRealpath(fd)
+		if err != nil {
+			unix.Close(fd)
+			return "", -1, err
+		}
+		if !isRelativeTo(realPath, basePath) {
+			unix.Close(fd)
+			return "", -1, ErrorEscapesBase{}
+		}
+
+		mountSource = procPath
+	}
+
+	return mountSource, prevFd, nil
+}
+
+func procFsFdPath(fd int) string {
+	pid := unix.Getpid()
+	return fmt.Sprintf("/proc/%d/fd/%d", pid, fd)
+}
+
+// getFdRealpath returns the real and procfs path for the given file descriptor.
+func getFdRealpath(fd int) (realPath string, procFsPath string, outErr error) {
+	procFd := procFsFdPath(fd)
+	realPath, err := os.Readlink(procFd)
```

Comment:
The `/proc/self/fd/[n]` link could be a link to a symlink if a symlink was opened with `open(O_PATH|O_NOFOLLOW)`. Since the function takes an arbitrary file descriptor, it is not safe in general to assume that readlink would produce the same result as realpath. Only the caller can know if that optimization is safe to use.